src/native/clr/host/bridge-processing.cc

```diff
@@ -0,0 +1,307 @@
+
+	abort_unless (GCUserPeer_class != nullptr && GCUserPeer_ctor != nullptr, "Failed to load mono.android.GCUserPeer!");
+}
+
+BridgeProcessing::BridgeProcessing (MarkCrossReferencesArgs *args) noexcept
+	: env{ OSBridge::ensure_jnienv () },
+	  cross_refs{ args }
+{
+}
+
+void BridgeProcessing::process () noexcept
+{
+	prepare_for_java_collection ();
+	GCBridge::trigger_java_gc (env);
+	cleanup_after_java_collection ();
+}
+
+void BridgeProcessing::prepare_for_java_collection () noexcept
+{
+	// Before looking at xrefs, scan the SCCs. During collection, an SCC has to behave like a
+	// single object. If the number of objects in the SCC is anything other than 1, the SCC
+	// must be doctored to mimic that one-object nature.
+	for (size_t i = 0; i < cross_refs->ComponentCount; i++) {
+		const StronglyConnectedComponent &scc = cross_refs->Components [i];
+
+		// Count > 1 case: The SCC contains many objects which must be collected as one.
+		// Solution: Make all objects within the SCC directly or indirectly reference each other
+		if (scc.Count > 1) {
+			add_circular_references (scc);
+		} else if (scc.Count == 0) {
+			// Some SCCs might have no IGCUserPeers associated with them, so we must create one
+			temporary_peers [i] = env->NewObject (GCUserPeer_class, GCUserPeer_ctor);
+		}
+	}
+
+	// Add the cross scc refs
+	for (size_t i = 0; i < cross_refs->CrossReferenceCount; i++) {
+		const ComponentCrossReference &xref = cross_refs->CrossReferences [i];
+		add_cross_reference (xref.SourceGroupIndex, xref.DestinationGroupIndex);
+	}
+
+	// With cross references processed, the temporary peer list can be released
+	for (const auto& [scc, temporary_peer] : temporary_peers) {
+		env->DeleteLocalRef (temporary_peer);
+	}
+
+	// Switch global to weak references
+	for (size_t i = 0; i < cross_refs->ComponentCount; i++) {
+		const StronglyConnectedComponent &scc = cross_refs->Components [i];
+		for (size_t j = 0; j < scc.Count; j++) {
+			take_weak_global_ref (scc.Contexts [j]);
+		}
+	}
+}
+
+CrossReferenceTarget BridgeProcessing::select_cross_reference_target (size_t scc_index) noexcept
+{
+	const StronglyConnectedComponent &scc = cross_refs->Components [scc_index];
+	if (scc.Count > 0) {
+		abort_unless (scc.Contexts [0] != nullptr, "SCC must have at least one context");
+		abort_unless (scc.Contexts [0]->control_block != nullptr, "SCC must have at least one context with valid control block");
+		return { .is_temporary_peer = false, .context = scc.Contexts [0] };
+	} else {
+		const auto temporary_peer = temporary_peers.find (scc_index);
+		abort_unless (temporary_peer != temporary_peers.end(), "Temporary peer must be found in the map");
+		return { .is_temporary_peer = true, .temporary_peer = temporary_peer->second };
+	}
+}
+
+void BridgeProcessing::add_circular_references (const StronglyConnectedComponent &scc) noexcept
+{
+	abort_unless (scc.Count > 1, "SCC must have at least two items to add inner references");
```

Comment:
This condition is checked at call site. Since this is a private function, add a comment for this precondition (above the function) and remove this line.